src/core/ext/filters/client_channel/resolver/fake/fake_resolver.c

```diff
@@ -151,29 +150,37 @@ void grpc_fake_resolver_response_generator_unref(
   }
 }
 
+typedef struct set_response_cb_arg {
+  grpc_fake_resolver_response_generator* generator;
+  grpc_channel_args* next_response;
+} set_response_cb_arg;
+
 static void set_response_cb(grpc_exec_ctx* exec_ctx, void* arg,
                             grpc_error* error) {
-  grpc_fake_resolver_response_generator* generator =
-      (grpc_fake_resolver_response_generator*)arg;
+  set_response_cb_arg* cb_arg = arg;
+  grpc_fake_resolver_response_generator* generator = cb_arg->generator;
   fake_resolver* r = generator->resolver;
   if (r->next_results != NULL) {
     grpc_channel_args_destroy(exec_ctx, r->next_results);
   }
-  r->next_results = generator->next_response;
+  r->next_results = cb_arg->next_response;
   if (r->results_upon_error != NULL) {
     grpc_channel_args_destroy(exec_ctx, r->results_upon_error);
   }
-  r->results_upon_error = grpc_channel_args_copy(generator->next_response);
+  r->results_upon_error = grpc_channel_args_copy(cb_arg->next_response);
+  gpr_free(cb_arg);
   fake_resolver_maybe_finish_next_locked(exec_ctx, r);
 }
 
 void grpc_fake_resolver_response_generator_set_response(
     grpc_exec_ctx* exec_ctx, grpc_fake_resolver_response_generator* generator,
     grpc_channel_args* next_response) {
   GPR_ASSERT(generator->resolver != NULL);
-  generator->next_response = grpc_channel_args_copy(next_response);
+  set_response_cb_arg* cb_arg = gpr_zalloc(sizeof(*cb_arg));
+  cb_arg->generator = generator;
+  cb_arg->next_response = grpc_channel_args_copy(next_response);
   GRPC_CLOSURE_SCHED(
-      exec_ctx, GRPC_CLOSURE_CREATE(set_response_cb, generator,
+      exec_ctx, GRPC_CLOSURE_CREATE(set_response_cb, cb_arg,
```

Comment:
Note that `GRPC_CLOSURE_CREATE()` does a malloc also.  Let's avoid doing two mallocs here by adding a `grpc_closure` member to `set_response_cb_arg` and using that here.

Reply:
Good idea. Done.